### PR TITLE
Port REACT_MODULE() from react-native-windows to react-native-macos (2/4)

### PR DESCRIPTION
**Pull request by ZihanChen-MSFT**
#### Please select one of the following
- [ ] I am removing an existing difference between facebook/react-native and microsoft/react-native-macos :thumbsup:
- [ ] I am cherry-picking a change from Facebook's react-native into microsoft/react-native-macos :thumbsup:
- [x] I am making a fix / change for the macOS implementation of react-native
- [ ] I am making a change required for Microsoft usage of react-native

## Summary

This is the second change for porting REACT_MODULE() from react-native-windows
It enables user to run a turbo module in C++ like this
```c++
REACT_MODULE(MyModule)
struct MyModule
{
    REACT_METHOD(MyMethod)
    void MyMethod(string a, int b, ReactPromise<double> result) noexcept
    {
        result.Reject("Not implemented");
    }
};
```

- This change moves some files from react-native-windows, they defines `REACT_MODULE` with a proper reflection system
- All copied files will eventually be downloaded from packages, the duplication is just a temporary solution for developing
- The turbo module written using `REACT_MODULE` is executable, but since threading supports is not added, the sample module only triggers an error.

## For Code Review

The original code is deeply coupled with WinRT,
I have no choice but to implement involved WinRT API in macOS,
here is two files that do this job

- **RNTester/RNTester-macOS/TurboModuleCxx/winrt/Windows.Foundation.h**
- **RNTester/RNTester-macOS/TurboModuleCxx/winrt/Microsoft.ReactNative.h**

For other files in **RNTester/RNTester-macOS/TurboModuleCxx**,
they are copied from react-native-windows,
which mean I am not going to fix issues about these files in this pull request
(because such changes are expected to be submitted to react-native-windows),
except for code in `#ifdef __APPLE__`.

As what is mentioned above,
eventually these files will be downloaded from a package,
so after a few pull requests,
they will be deleted from this repo.

Having duplicated files here is a temporary solution for development convenience.

## About Future Pull Requests

More works are needed to make this scenario completes.
In the next pull request,
I will add threading supports so that user can navigate between UI thread and JS thread.

## Changelog

[macOS] [Fixed] - Port REACT_MODULE() from react-native-windows to react-native-macos (2/4)

## Test Plan

- Manually tested

###### Microsoft Reviewers: [Open in CodeFlow](http://wpcp.azurewebsites.net/CodeFlowProtocolProxy2.php?pullrequest=https://github.com/microsoft/react-native-macos/pull/599)

### CODE REVIEW

File: RNTester/RNTester-macOS/TurboModuleCxx/TurboModulesProvider.cpp [hunk 1]
```diff
@@ -0,0 +1,304 @@
...
+                  if (method.ReturnType == MethodReturnType::Callback) {
+                    resolveFunction = {runtime, args[count - 1]};
+                  } else {
+                    resolveFunction = {runtime, args[count - 2]};
+                    rejectFunction = {runtime, args[count - 1]};
+                  }
+
+                  auto makeCallback = [&runtime](
+                                          const facebook::jsi::Value &callbackValue) noexcept->MethodResultCallback {
+                    // workaround: xcode doesn't accept a captured value with only rvalue copy constructor
```
Comment (alloy) [inline RNTester/RNTester-macOS/TurboModuleCxx/TurboModulesProvider.cpp:191]: This is a 10.13 deployment target workaround, right? If so, would be good to reword this to clarify the actual problem.

File: RNTester/RNTester-macOS/TurboModuleCxx/winrt/Microsoft.ReactNative.h [hunk 2]
```diff
@@ -5,6 +5,12 @@
 namespace winrt::Microsoft::ReactNative
 {
 
+template <typename T>
+decltype(auto) ReadOptional(std::optional<T> &opt)
+{
+  return opt.__get();
+}
```
Comment (alloy) [inline RNTester/RNTester-macOS/TurboModuleCxx/winrt/Microsoft.ReactNative.h:12]: Does this need a `#ifndef __APPLE__` guard as well?

**Review by alloy — APPROVED** [review]
Some minor comments 👍

File: RNTester/NativeModuleExample/ScreenshotMacOS.mm [hunk 3]
```diff
@@ -141,16 +142,71 @@ void TakeScreenshot(
...
+  std::shared_ptr<winrt::Microsoft::ReactNative::TurboModulesProvider> provider;
+  winrt::Microsoft::ReactNative::IReactContext reactContext;
+}
 
 - (std::shared_ptr<facebook::react::TurboModule>)
   getTurboModule:(const std::string &)name
   jsInvoker:(std::shared_ptr<facebook::react::CallInvoker>)jsInvoker
 {
-  if (name == "ScreenshotManager")
+  if (!provider)
```
[the remark this answers is not in the document]
Comment (ZihanChen-MSFT) [reply RNTester/NativeModuleExample/ScreenshotMacOS.mm:194]: I believe so, when debugging this function I found this is called for all used turbo modules name in javascript at launch time.

File: RNTester/RNTester-macOS/TurboModuleCxx/winrt/Microsoft.ReactNative.h (hunk 2 above)
Comment (ZihanChen-MSFT) [reply RNTester/RNTester-macOS/TurboModuleCxx/winrt/Microsoft.ReactNative.h:12]: No, this file only exist in react-native-macos.

File: RNTester/NativeModuleExample/ScreenshotMacOS.mm [hunk 4]
```diff
@@ -141,16 +142,71 @@ void TakeScreenshot(
...
+      L"ScreenshotManager",
+      winrt::Microsoft::ReactNative::MakeModuleProvider<ScreenshotManagerCxx>()
+      );
   }
+  return provider->getModule(name, jsInvoker);
+  // if (name == "ScreenshotManager")
+  // {
+  //   return std::make_shared<ScreenshotManagerTurboModule>(jsInvoker);
+  // }
   return nullptr;
```
[the remark this answers is not in the document]
Comment (ZihanChen-MSFT) [reply RNTester/NativeModuleExample/ScreenshotMacOS.mm:210]: This will be deleted in the next pull request, I believe `ScreenshotManagerTurboModule` is able to be completely converted to `ScreenManagerCxx` at that moment.
